man/atopcat.1
.TH ATOPCAT 1 "January 2024" "Linux"
.SH NAME
.B atopcat
- concatenate raw log files to stdout
.SH SYNOPSIS
.P
.B atopcat [-dv] rawfile [rawfile]...
.P
.SH DESCRIPTION
The program
.I atopcat
can be used to concatenate several raw log files into one stream (stdout).
In this way, raw log files can be merged into one larger file by redirecting
stdout to a file. Alternatively, merged data from several raw log files
can be transferred directly into
.I atop
or
.I atopsar
via a pipe.

Options:
.PP
.TP 5
.B -d
dry-run: read logfile(s) but do not generate output on stdout
.PP
.TP 5
.B -v
verbose: print one line per sample containing date/time, interval length
in seconds, compressed length of the system-level information,
compressed length of the process-level information,
compressed length of the cgroup-level information, and
compressed length of the PID list.
.SH EXAMPLES
Concatenate the raw log files of five contiguous working days,
write it into a new raw log file for that week and
view that week interactively:
.PP
.TP 12
.B \  atopcat /var/log/atop/atop_2024021[0-4] > week_2024_7
.TP 12
.B \  atop -r week_2024_7
.PP
Concatenate the raw log files of a week and view that week interactively
(since
.I atop
reads from a pipe, previous intervals can not be retrieved while viewing): 
.PP
.TP 12
.B \  atopcat /var/log/atop/atop_2024021[0-6] | atop -r -
.PP
Concatenate all raw log files of January 2024 and generate parsable
output about the CPU utilization:
.PP
.TP 12
.B \  atopcat /var/log/atop/atop_202401?? | atop -r -  -PCPU
.PP
Concatenate the daily raw log files of February 3 and 4,
and generate a report about memory utilization from 14:00h on the first day
till 11:00h on the second day:
.PP
.TP 12
.B \  atopcat /var/log/atop/atop_2024020[34] |
.B \  atopsar -m  -r -  -b 202402031400  -e 202402041100
.PP
Repair a raw log file from which the last interval has not been
completely written (e.g. if you intend to expand the file with new samples):
.PP
.TP 12
.B \  atopcat /var/log/atop/atop_20240303 > /tmp/repaired
.PP
In the latter case,
.B atopcat
reports that the input file is incomplete and stops after the last consistent
sample.
.SH SEE ALSO
.B atop(1),
.B atopsar(1),
.B atophide(1),
.B atopconvert(1)
.br
.B https://www.atoptool.nl
.SH AUTHOR
Gerlof Langeveld (gerlof.langeveld@atoptool.nl)


ChangeLog
2010-11-30  Gerlof Langeveld  <gerlof.langeveld@atoptool.nl>

        * showprocs.c
		Dash (-) added in the column PPID for exited processes.

2010-11-26  Gerlof Langeveld  <gerlof.langeveld@atoptool.nl>

        * atopsar.c
		When atopsar is called to print more than one report
		from a raw logfile with a specified begin time, all
		reports should be printed from that begin time (not
		only the first).

2010-11-19  Gerlof Langeveld  <gerlof.langeveld@atoptool.nl>

        * photosyst.c
		Support for virtual disk vd... (kvm).

2010-11-17  Gerlof Langeveld  <gerlof.langeveld@atoptool.nl>

        * rawlog.c
		The flag -r followed by exactly 8 'y' characters is not
		considered as 8 days ago, but as a literal filename.

2010-11-14  Gerlof Langeveld  <gerlof.langeveld@atoptool.nl>

        * photosyst.c
		Overflow of file descriptors because /proc/cpuinfo was
		opened for every interval, but not closed any more.
		Credits: Dardo Kleiner

2010-11-12  Gerlof Langeveld  <gerlof.langeveld@atoptool.nl>

        * various.c
		Show all parts of timestamp in header line, even when zero.

2010-11-12  Gerlof Langeveld  <gerlof.langeveld@atoptool.nl>

        * showsys.c, showprocs.c
		Sometimes segmentation-fault on particular CPU-types
		due to memcpy i.s.o. memmove when moving memory in overlap.

2010-10-25  Gerlof Langeveld  <gerlof.langeveld@atoptool.nl>

        * showgeneric.c
		When the number of lines is too small for the system-level
		lines, limit the number of variable resources automatically
		to a minimum.

2010-10-23  Gerlof Langeveld  <gerlof.langeveld@atoptool.nl>

        * atopsar.c
		Optimized code to show total number of running/sleeping
		threads.

2010-10-23  Gerlof Langeveld  <gerlof.langeveld@atoptool.nl>

        * photosyst.c
		Add support for mmcblk disks (MMC/SD cardreaders)    
		Credits: Anssi Hannula

2010-10-23  JC van Winkel  <jc@ATComputing.nl>

        * atop.c, atopsar.c, deviate.c, parseable.c, rawlog.cs, showgeneric.c,
	showlinux.c, showsys.c
		Add counters for total number of running and sleeping (S and
		D) threads.

2010-05-18  JC van Winkel  <jc@ATComputing.nl>

        * various.c, showsys.c, deviate.c, photosyst.c, parseable.c, showlinux.c
		Introduce CPU frequency and scaling.

2010-04-28  Gerlof Langeveld  <gerlof.langeveld@atoptool.nl>

        * various.c
		Cast value larger than 4GB to long long.

2010-04-23  Gerlof Langeveld  <gerlof.langeveld@atoptool.nl>

        * showprocs.c
		Added special routined for uid/gid not available for exited
		processes.	

2010-04-23  Gerlof Langeveld  <gerlof.langeveld@atoptool.nl>

        * showgeneric.c, atop.c
		Version (flag -V) handled earlier after startup.

2010-04-23  JC van Winkel  <jc@ATComputing.nl>

        * showsys.c
		Field 'avque' modified to 'avq' to be able to show higher
		values (especially on LVM-level).

2010-04-23  Gerlof Langeveld  <gerlof.langeveld@atoptool.nl>

        * showgeneric.c
		Proper sorting of processes when switching from single
		process view to cumulative view (key 'u' or 'p') and vice
		versa.

2010-04-17  JC van Winkel  <jc@ATComputing.nl>

        * atop.c, showgeneric.c, showlinux.c
		Allow modifying the layout of the columns in the system
		lines via ~/.atoprc and /etc/atoprc

2010-04-16  Gerlof Langeveld  <gerlof.langeveld@atoptool.nl>

        * atop.c, rawlog.c
  		Automatically start another version of atop if the logfile
		to be read has not been created by the current version.

2010-03-26  Gerlof Langeveld  <gerlof.langeveld@atoptool.nl>

        * various.c
  		Introduced unit of Tbytes for memory-usage.

2010-03-19  Gerlof Langeveld  <gerlof.langeveld@atoptool.nl>

        * 45atoppm
  		Add script to handle power management for laptops, i.e.
		to take an extra sample just before hibernate/suspend and
		to restart atop in case of thaw/resume.

2010-03-16  Gerlof Langeveld  <gerlof.langeveld@atoptool.nl>

        * showgeneric.c, showlinux.c
  		Program and user selection can be combined with program and
		user accumulation.

2010-03-16  Gerlof Langeveld  <gerlof.langeveld@atoptool.nl>

        * showgeneric.c
  		Show in header-line if user selection (U) and
		program selection (P) is active.

2010-03-16  Gerlof Langeveld  <gerlof.langeveld@atoptool.nl>

        * showgeneric.c
  		Performance improvement - only sort system-resources once per
		interval.

2010-03-05  Gerlof Langeveld  <gerlof.langeveld@atoptool.nl>

        * photosyst.c
  		Add recognition of disk type "/dev/fio...".

2010-03-05  Gerlof Langeveld  <gerlof.langeveld@atoptool.nl>

        * photosyst.c, atop.c, atopsar.c, deviate.c, parseable.c,
	  showgeneric.c, showlinux.c, showsys.c
  		Major change to support I/O-statistics on LVM and MD level.

2010-03-05  JC van Winkel  <jc@ATComputing.nl>

        * showsys.c
  		CPU-number was wrong in the sorted per-cpu statistics.

2010-03-05  Gerlof Langeveld  <gerlof.langeveld@atoptool.nl>

        * rawlog.c
  		Struct stat size did not fit in short any more (modified to
		int) in the header of the raw file.

2010-02-16  Gerlof Langeveld  <gerlof.langeveld@atoptool.nl>

        * showgeneric.c
  		Bug-solution for systems with long uptime (float exception);
		increased buffer-length for call to val2elapstr.

2010-01-18  Gerlof Langeveld  <gerlof.langeveld@atoptool.nl>

        * showlinux.c
  		Modified priorities for system-level columns.

2010-01-16  Gerlof Langeveld  <gerlof.langeveld@atoptool.nl>

        * showprocs.c, showlinux.c
  		Minor changes in output-specification.

2010-01-16  JC van Winkel  <jc@ATComputing.nl>

        * showprocs.c, showlinux.c
  		Corrected counters for patched kernels (JC van Winkel).

2010-01-08  Gerlof Langeveld  <gerlof.langeveld@atoptool.nl>

        * parseable.c
  		Added label RESET in case of a sample with values since boot.

2010-01-08  Gerlof Langeveld  <gerlof.langeveld@atoptool.nl>

        * showprocs.c
  		Added policies batch, iso and idle for scheduling class.


atopcat.c
/*
** ATOP - System & Process Monitor
**
** The program 'atop' offers the possibility to view the activity of 
** the system on system-level as well as process-level.
**
** This program concatenates several raw logfiles into one output stream,
** to be stored as new file or to be passed via a pipe to atop/atopsar directly.
** ==========================================================================
** Author:      Gerlof Langeveld
** E-mail:      gerlof.langeveld@atoptool.nl
** Initial:     March 2020
** --------------------------------------------------------------------------
** Copyright (C) 2020 Gerlof Langeveld
**
** This program is free software; you can redistribute it and/or modify it
** under the terms of the GNU General Public License as published by the
** Free Software Foundation; either version 2, or (at your option) any
** later version.
**
** This program is distributed in the hope that it will be useful, but
** WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
** See the GNU General Public License for more details.
**
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
** --------------------------------------------------------------------------
*/
#include <sys/types.h>
#include <sys/param.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <time.h>
#include <stdio.h>
#include <errno.h>
#include <stdlib.h>
#include <sys/utsname.h>
#include <unistd.h>

#include "atop.h"
#include "rawlog.h"

char	*convepoch(time_t);
void	prusage(char *);

int
main(int argc, char *argv[])
{
	int			i, fd, n, c;
	int			firstfile, beverbose=0, dryrun=0;
	struct rawheader	rh;
	struct rawrecord	rr;
	char			*infile, *sstat, *pstat, *cstat, *istat;
	unsigned short		aversion;

	// verify the command line arguments: input filename(s)
	//
	if (argc < 2)
		prusage(argv[0]);

	while ((c = getopt(argc, argv, "?hvd")) != EOF)
	{
		switch (c)
		{
		   case '?':            	// usage wanted?
		   case 'h':            	// usage wanted?
			prusage(argv[0]);
			break;

		   case 'v': 			// verbosity wanted?
			beverbose = 1;
			break;

		   case 'd': 			// dry run wanted?
			dryrun = 1;
			break;

		   default:
			prusage(argv[0]);
		}
	}

	if ( isatty(fileno(stdout)) && !dryrun)
	{
		fprintf(stderr,
			"this program produces binary output on stdout "
			"that should be redirected\nto a file or pipe!\n");
		exit(1);
	}

	// open all input files one-by-one
	//
	for (i=optind, firstfile=1; i < argc; i++, firstfile=0)
	{
		infile = argv[i];

		// open raw file for reading
		//
		if ( (fd = open(infile, O_RDONLY)) == -1)
		{
			fprintf(stderr, "%s - ", infile);
			perror("open for reading");
			exit(2);
		}

		// read the raw header
		//
		if ( read(fd, &rh, sizeof rh) < sizeof rh)
		{
			fprintf(stderr, "%s: cannot read raw header\n", infile);
			close(fd);
			exit(3);
		}

		// verify if this is a correct rawlog file
		//
		if (rh.magic != MYMAGIC)
		{
			fprintf(stderr, "%s: not a valid rawlog file "
				"(wrong magic number)\n", infile);
			close(fd);
			exit(4);
		}

		// only for the first file, store the version number and write
		// the raw header for the entire stream
		//
		// for the next files, be sure that the version is the same
		// as the first file
		//
		if (firstfile)
		{
			aversion = rh.aversion;

			if (!dryrun)
			{
				if ( write(1, &rh, sizeof rh) < sizeof rh)
				{
					fprintf(stderr,
						"can not write raw header\n");
					exit(10);
				}
			}

			if (beverbose)
			{
				fprintf(stderr,
				     "Logs created by atop version %d.%d\n\n", 
                        	     (rh.aversion >> 8) & 0x7f,
				      rh.aversion & 0xff);

				fprintf(stderr, "%-10s %-8s %12s  %8s  %9s  %8s %8s\n",
					"date", "time", "interval",
					"comprsys", "comprproc", "comprcgr", "comppids");
			}
					
		}
		else	// subsequent file
		{
			if (aversion != rh.aversion)
			{
				fprintf(stderr,
					"Version of file %s is unequal to "
					"version of first file\n", infile);
				close(fd);
				exit(5);
			}
		}

		// read every raw record followed by the compressed
		// system-level stats, process-level stats,
		// cgroup-level stats and pidlist.
		//
		while ( read(fd, &rr, sizeof rr) == sizeof rr )
		{
			if (beverbose)
			{
				fprintf(stderr, "%19s %12u  %8u  %9u  %8u %8u  %s\n",
					convepoch(rr.curtime),
					rr.interval, rr.scomplen, rr.pcomplen,
					rr.ccomplen, rr.icomplen,
					rr.flags&RRBOOT ? "boot" : "");
			}

			// dynamically allocate space to read stats
			// 
			if ( (sstat = malloc(rr.scomplen)) == NULL)
			{
				fprintf(stderr, "malloc failed for sstat\n");
				exit(7);
			}

			if ( (pstat = malloc(rr.pcomplen)) == NULL)
			{
				fprintf(stderr, "malloc failed for pstat\n");
				exit(7);
			}

			if ( (cstat = malloc(rr.ccomplen)) == NULL)
			{
				fprintf(stderr, "malloc failed for cstat\n");
				exit(7);
			}

			if ( (istat = malloc(rr.icomplen)) == NULL)
			{
				fprintf(stderr, "malloc failed for istat\n");
				exit(7);
			}

			// read system-level stats
			// 
			if ((n = read(fd, sstat, rr.scomplen)) != rr.scomplen)
			{
				if (n == -1)
				{
					fprintf(stderr, "read file %s", infile);
					perror("");
					exit(8);
				}
				else
				{
					fprintf(stderr,
					     "file %s incomplete!\n", infile);

					free(sstat);
					free(pstat);
					free(cstat);
					free(istat);
					break;
				}
			}

			// read process-level stats
			// 
			if ((n = read(fd, pstat, rr.pcomplen)) != rr.pcomplen)
			{
				if (n == -1)
				{
					fprintf(stderr, "read file %s", infile);
					perror("");
					exit(8);
				}
				else
				{
					fprintf(stderr,
					     "file %s incomplete!\n", infile);

					free(sstat);
					free(pstat);
					free(cstat);
					free(istat);
					break;
				}
			}

			// read cgroup-level stats
			// 
			if ((n = read(fd, cstat, rr.ccomplen)) != rr.ccomplen)
			{
				if (n == -1)
				{
					fprintf(stderr, "read file %s", infile);
					perror("");
					exit(8);
				}
				else
				{
					fprintf(stderr,
					     "file %s incomplete!\n", infile);

					free(sstat);
					free(pstat);
					free(cstat);
					free(istat);
					break;
				}
			}

			// read compressed pidlist
			// 
			if ((n = read(fd, istat, rr.icomplen)) != rr.icomplen)
			{
				if (n == -1)
				{
					fprintf(stderr, "read file %s", infile);
					perror("");
					exit(8);
				}
				else
				{
					fprintf(stderr,
					     "file %s incomplete!\n", infile);

					free(sstat);
					free(pstat);
					free(cstat);
					free(istat);
					break;
				}
			}


			if (!dryrun)
			{
				// write raw record followed by the compressed
				// system-level stats, process-level stats,
				// cgroup-level stats and pidlist
				//
				if ( write(1, &rr, sizeof rr) < sizeof rr)
				{
					fprintf(stderr, "can not write raw record\n");
					exit(11);
				}

				if ( write(1, sstat, rr.scomplen) < rr.scomplen)
				{
					fprintf(stderr, "can not write sstat\n");
					exit(11);
				}

				if ( write(1, pstat, rr.pcomplen) < rr.pcomplen)
				{
					fprintf(stderr, "can not write pstat\n");
					exit(11);
				}

				if ( write(1, cstat, rr.ccomplen) < rr.ccomplen)
				{
					fprintf(stderr, "can not write cstat\n");
					exit(11);
				}

				if ( write(1, istat, rr.icomplen) < rr.icomplen)
				{
					fprintf(stderr, "can not write istat\n");
					exit(11);
				}
			}

			// free dynamically allocated buffers
			//
			free(sstat);
			free(pstat);
			free(cstat);
			free(istat);
		}

		close(fd);
	}

	return 0;
}

// Function to convert an epoch time to date-time format
//
char *
convepoch(time_t utime)
{
	struct tm       *tt;
	static char	datetime[64];
	

	tt = localtime(&utime);

	sprintf(datetime, "%04d/%02d/%02d %02d:%02d:%02d",
                tt->tm_year+1900, tt->tm_mon+1, tt->tm_mday,
		tt->tm_hour, tt->tm_min, tt->tm_sec);

        return datetime;
}

// Function that shows the usage message
//
void
prusage(char *name)
{
	fprintf(stderr, "Usage: %s [-dv] rawfile [rawfile]...\n", name);
	fprintf(stderr, "\t-d\tdry run (no raw output generated)\n");
	fprintf(stderr, "\t-v\tbe verbose\n");
	exit(1);
}
